File: src/main/java/dev/liquid/osamaclient/OsamaClient.java
package dev.liquid.osamaclient;

import dev.liquid.osamaclient.config.OsamaClientConfig;
import net.minecraft.client.Minecraft;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;

import static dev.liquid.osamaclient.util.ClassReference.featureManager;

@Mod(modid = OsamaClient.MODID, name = OsamaClient.NAME, version = OsamaClient.VERSION)
public class OsamaClient {
  public static final String MODID = "osamaclient";
  public static final String NAME = "osamaClient";
  public static final String VERSION = "%%VERSION%%";
  public static OsamaClientConfig config;
  @Mod.Instance(MODID)
  private static OsamaClient instance;
  public static OsamaClient getInstance() {
    if(instance == null){
      instance = new OsamaClient();
    }
    return instance;
  }
  @Mod.EventHandler
  private void onInit(FMLInitializationEvent event) {
    config = new OsamaClientConfig();

    featureManager.loadFeatures().forEach(MinecraftForge.EVENT_BUS::register);
  }
}

File: src/main/kotlin/dev/liquid/osamaclient/OsamaClient.kt
package dev.liquid.osamaclient

import cc.polyfrost.oneconfig.utils.commands.CommandManager
import dev.liquid.osamaclient.command.Set
import dev.liquid.osamaclient.config.OsamaClientConfig
import dev.liquid.osamaclient.feature.FeatureManager
import dev.liquid.osamaclient.util.cache.CachedWorld
import dev.liquid.osamaclient.util.cache.WorldScanner
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.event.FMLInitializationEvent

@Mod(modid = "osamaclient", name = "osamaClient", version = "%%VERSION%%")
class OsamaClient {

  // Singleton Getter
  companion object {
    @Mod.Instance("osamaclient")
    lateinit var INSTANCE: OsamaClient private set
  }

  lateinit var config: OsamaClientConfig private set
  lateinit var cachedWorld: CachedWorld private set
  lateinit var worldScanner: WorldScanner private set

  @Mod.EventHandler
  private fun onInit(event: FMLInitializationEvent) {
    config = OsamaClientConfig()
    cachedWorld = CachedWorld()
    worldScanner = WorldScanner()
    val cmd = Set()

    FeatureManager.getInstance().loadFeatures().forEach { feature -> MinecraftForge.EVENT_BUS.register(feature) }
    MinecraftForge.EVENT_BUS.register(cmd)
    MinecraftForge.EVENT_BUS.register(worldScanner)
    CommandManager.register(cmd)
  }
}
